Core/NoOpRunner.Core/Shapes/Platform.cs

```diff
@@ -1,61 +1,95 @@
 ﻿using NoOpRunner.Core.Entities;
 using NoOpRunner.Core.Enums;
-using System;
+using System.Collections.Generic;
 
 namespace NoOpRunner.Core.Shapes
 {
     public class Platform : BaseShape
     {
-        private Random random = new Random();
         public int bottomPosY { get; set; } // Bottom of x platform
         public int topPosY { get; set; } // Max height of x platform
 
         public int lastPosX = 0,
                    lastPosY = 0;
 
+        private List<int> xCoords = new List<int>();
+        private List<int> yCoords = new List<int>();
+
         public Platform(int centerPosX, int centerPosY, int bottomPosY, int topPosY) : base(centerPosX, centerPosY)
         {
             while (lastPosX < 28) // TODO
             {
-                int blockLength = randomLength(4);
-                int blockHeight = randomHeight(4); // Lower than jump height
+                int blockLength = RandomLength(4);
+                int blockHeight = RandomHeight(4); // Lower than jump height
 
-                MapShapeX(lastPosX, lastPosY, blockLength, Color.Red);
+                AddShape(true, lastPosX, lastPosY, blockLength);
                 lastPosX += blockLength;
+
                 if(blockHeight < 0)
                 {
-                    MapShapeY(lastPosX, 1, lastPosY, Color.Red);
+                    AddShape(false, lastPosX, 1, lastPosY);
                     lastPosY = 0;
                 } else
                 {
                     if (lastPosY + blockHeight > topPosY)
                     {
-                        MapShapeY(lastPosX, lastPosY, (topPosY - lastPosY), Color.Red);
+                        AddShape(false, lastPosX, lastPosY, (topPosY - lastPosY) < bottomPosY ? bottomPosY : (topPosY - lastPosY));
                         lastPosY = topPosY;
                     } else
                     {
-                        MapShapeY(lastPosX, lastPosY, blockHeight, Color.Red);
+                        AddShape(false, lastPosX, lastPosY, blockHeight);
                         lastPosY = lastPosY + blockHeight;
                     }
                 }
             }
         }
 
+        public void AddShape(bool horizontal, int posX, int posY, int blockLength)
```

Comment:
This looks like it should be a BaseShape method as protected, and the MapShapeX/Y methods as private. Same for GetCoordsX, GetCoordsY.

Reply:
@t-eliks Can't quite understand this comment. AddShape is a custom function inside Platform.cs to put cordinates into list easily instead of using the same code in many rows as the constructor is already quite complex. GetCoordsX and GetCoordsY can't be private as PowerUp spawn location depends on these functions.

Reply:
The way you implemented the logic right now, means power ups can only spawn on platforms and no other shape. You should move the AddShape method to a base class as protected, and GetCoords methods as public methods in that base class, so that all shapes can have powerups spawn on them.

Reply:
Done

Reply:
Updated BaseShape.cs. There is no need to move AddShape function to BaseShape class as that function only saves some code rows inside Platform class.